ultralytics/solutions/dwell_time_analyzer.py

```diff
@@ -0,0 +1,475 @@
+        self.analytics_text_color: Tuple[int, int, int] = self.CFG.get("analytics_text_color", (0, 0, 0))
+        self.zones: Dict[str, List[Tuple[int, int]]] = self.CFG.get("zones", {})
+        self.classes: Optional[List[int]] = self.CFG.get("classes", None)  # If provided, only track these classes
+
+        # Optional features
+        self.enable_funnel: bool = self.CFG["enable_funnel"]
+        self.funnel_stages: Optional[Union[List[str], Tuple[str, ...]]] = self.CFG["funnel_stages"]
+        self.enable_avg_dwell: bool = self.CFG["enable_avg_dwell"]
+        self.detect_mode: str = self.CFG["detect_mode"]  # "all_frames" or "enter_zones"
+
+        # Validate detect_mode
+        if self.detect_mode not in ["all_frames", "enter_zones"]:
+            raise ValueError("detect_mode must be either 'all_frames' or 'enter_zones'.")
+
+        # If no zones provided, interactively select them
+        if not self.zones or len(self.zones) == 0:
+            self.zones = self.select_zones()
+
+        self.zone_polygons: Dict[str, Union[self.Polygon, self.LineString]] = {}
+        for zone_name, coords in self.zones.items():
+            if len(coords) >= 3:
+                self.zone_polygons[zone_name] = self.Polygon(coords)
+            else:
+                self.zone_polygons[zone_name] = self.LineString(coords)
+
+        # Dynamically determine funnel stages if not provided and funnel is enabled
+        if self.enable_funnel:
+            if not self.funnel_stages or len(self.funnel_stages) < 2:
+                # Use all zones as funnel stages if possible
+                zone_names = list(self.zones.keys())
+                if len(zone_names) >= 2:
+                    self.funnel_stages = tuple(zone_names)  # All zones form the funnel
+                else:
+                    # Not enough zones for a funnel
+                    self.funnel_stages = None
+
+        # Tracking variables
+        self.object_current_zone: Dict[int, str] = {}  # track_id -> zone_name
+        self.object_zone_entry_time: Dict[Tuple[int, str], float] = {}  # (track_id, zone_name) -> entry_time
+        self.object_zone_sequence: defaultdict = defaultdict(list)  # track_id -> [(zone_name, entry_time, exit_time)]
+        self.frame_number: int = 0
+
+    def select_zones(self) -> Dict[str, List[Tuple[int, int]]]:
+        """
+        Interactively selects zones on the first frame of the source video.
+
+        Returns:
+            Dict[str, List[Tuple[int, int]]]: A dictionary mapping zone names to their polygon points.
+
+        Raises:
+            ValueError: If no video source is provided in the configuration.
+            IOError: If the video source cannot be opened or the first frame cannot be read.
+        """
+        if not self.CFG.get("source"):
+            raise ValueError("No source video provided. Cannot interactively select zones.")
+
+        cap = cv2.VideoCapture(self.CFG["source"])
```

Comment:
@Muhammad-237 Hi, why we are reading the video file again here? Please follow the format we follow for other solutions, we don't read the video file inside any solution, the user calls it via VideoCapture and then passes each frame to the solutions class for processing! Thank you!

Reply:
yes, because i have read the parking management code where we upload an image for drawing zones. so there might a be a case when user don't have an image so in source i added video which is necessary when user don't provide hardcoded region he/she can interactively draw from their source.